plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -454,4 +430,44 @@ public static List<ParquetReaderColumn> createParquetReaderColumns(List<HiveColu
                 })
                 .collect(toImmutableList());
     }
+
+    private static Optional<Type> getBaseColumnType(HiveColumnHandle column, MessageType messageType, boolean useParquetColumnNames)
+    {
+        if (useParquetColumnNames) {
+            return Optional.ofNullable(getParquetTypeByName(column.getBaseColumnName(), messageType));
+        }
+        if (column.getBaseHiveColumnIndex() < messageType.getFieldCount()) {
+            return Optional.of(messageType.getType(column.getBaseHiveColumnIndex()));
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Dereferencing base parquet type based on projection info's dereference names.
+     * For example, when dereferencing baseType(level1Field0, level1Field1, Level1Field2(Level2Field0, Level2Field1))
+     * with a projection info's dereferenceNames list as (basetype, Level1Field2, Level2Field1).
+     * It would return a list of parquet types in the order of (level1Field2, Level2Field1).*
+     *
+     * @return child fields on each level of dereferencing. Return emptyList when failed to do lookup.
+     */
+    private static List<Type> dereferenceSubFieldTypes(GroupType baseType, HiveColumnProjectionInfo projectionInfo)
```

Comment:
Maybe `Optional<List<Type>>` and use empty for the lookup fails

Reply:
thanks, done!